web3/providers/persistent/subscription_manager.py

```diff
@@ -0,0 +1,274 @@
+import asyncio
+import logging
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Dict,
+    Iterator,
+    List,
+    Optional,
+    Sequence,
+    Union,
+    cast,
+    overload,
+)
+
+from eth_typing import (
+    HexStr,
+)
+
+from web3.exceptions import (
+    SubscriptionProcessingFinished,
+    TaskNotRunning,
+    Web3TypeError,
+    Web3ValueError,
+)
+from web3.types import (
+    FormattedEthSubscriptionResponse,
+    RPCResponse,
+)
+from web3.utils.subscriptions import (
+    EthSubscription,
+    EthSubscriptionContext,
+)
+
+if TYPE_CHECKING:
+    from web3 import AsyncWeb3  # noqa: F401
+    from web3.providers.persistent import (  # noqa: F401
+        PersistentConnectionProvider,
+        RequestProcessor,
+    )
+
+
+class SubscriptionContainer:
```

Comment:
more nits: Do you think this `SubscriptionContainer` class deserves its own file?

Reply:
I could go either way. Do you? 😆 I don't think this file is too large but happy to abstract it out. Thoughts?

Reply:
I could go either way too. I think I'd slightly prefer it, but not a strong opinion and we can always do it later if it gets unwieldy :)